### ISSUE TITLE
Floating point inconsistencies when applying aggregation function to result of non-integer functions (data flow vs. constant folding)

### ISSUE THREAD

**Issue by nrainer-materialize** [MaterializeInc/materialize#19592]
### What version of Materialize are you using?

v0.55.0 (448db5f3c)

### What is the issue?

Expressions like `sum(cot(int4_one))` or `sum(radians(uint2_val))` produce a different result depending on the evaluation strategy (data-flow rendering vs. constant folding). The involved functions seem to <s>(always?)</s> be trigonometric ones.
This was identified with the output consistency test framework when adding nesting functionality (https://github.com/MaterializeInc/materialize/pull/19523).

#### 1) sum of cot
* Expression: `sum(cot(int4_one))`
* Data-flow rendering: `0.6420925855636597`
* Constant folding: `0.6420926159343306`

<details>
  <summary>SQL</summary>

```
-- Setup for evaluation strategy 'Dataflow rendering':
DROP TABLE IF EXISTS t_dfr_horiz;
CREATE TABLE t_dfr_horiz (int4_one INT4);
INSERT INTO t_dfr_horiz VALUES (1::INT4);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_horiz (int4_one)
 AS SELECT 1::INT4;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
 sum(cot(int4_one))
FROM
  t_dfr_horiz;
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
 sum(cot(int4_one))
FROM
  v_ctf_horiz;
```

</details>

#### 2) sum of radians
* Expression: `sum(radians(uint2_val))`
* Data-flow rendering: `0.017453253269195557`
* Constant folding: `0.017453292519943295`

<details>
  <summary>SQL</summary>

```
-- Setup for evaluation strategy 'Dataflow rendering':
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, uint2_val UINT2);
INSERT INTO t_dfr_vert VALUES (1, 0::UINT2);
INSERT INTO t_dfr_vert VALUES (5, 1::UINT2);
INSERT INTO t_dfr_vert VALUES (7, 0::UINT2);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, uint2_val)
 AS SELECT 1, 0::UINT2
    UNION SELECT 5, 1::UINT2
    UNION SELECT 7, 0::UINT2;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  sum(radians(uint2_val))
FROM
  t_dfr_vert
WHERE row_index IN (1, 5, 7);
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  sum(radians(uint2_val))
FROM
  v_ctf_vert
WHERE row_index IN (1, 5, 7);
```

</details>

#### 3) stddev_samp on asin
* Expression: `stddev_samp(asin(uint4_val))`
* Data-flow rendering: `0.9068996831246972`
* Constant folding: `0.9068996821171088`

<details>
  <summary>SQL</summary>

```
-- Setup for evaluation strategy 'Dataflow rendering':
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, uint4_val UINT4);
INSERT INTO t_dfr_vert VALUES (1, 0::UINT4);
INSERT INTO t_dfr_vert VALUES (4, 0::UINT4);
INSERT INTO t_dfr_vert VALUES (8, 1::UINT4);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, uint4_val)
 AS SELECT 1, 0::UINT4
    UNION SELECT 4, 0::UINT4
    UNION SELECT 8, 1::UINT4;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  stddev_samp(asin(uint4_val))
FROM
  t_dfr_vert
WHERE row_index IN (8, 1, 4);
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  stddev_samp(asin(uint4_val))
FROM
  v_ctf_vert
WHERE row_index IN (8, 1, 4);
```

</details>

#### 4) avg on acos
* Expression: `avg(acos(int8_val))`
* Data-flow rendering: `0.7853981554508209`
* Constant folding: `0.7853981633974483`

<details>
  <summary>SQL</summary>

```
-- Setup for evaluation strategy 'Dataflow rendering':
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, int8_val INT8);
INSERT INTO t_dfr_vert VALUES (2, 1::INT8);
INSERT INTO t_dfr_vert VALUES (9, 0::INT8);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, int8_val)
 AS SELECT 2, 1::INT8
    UNION SELECT 9, 0::INT8;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  avg(acos(int8_val))
FROM
  t_dfr_vert
WHERE row_index IN (9, 2);
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  avg(acos(int8_val))
FROM
  v_ctf_vert
WHERE row_index IN (9, 2);
```

</details>

#### 5) avg on exp
* Expression: `avg(exp(uint4_val))`
* Data-flow rendering: `1.8591409027576447`
* Constant folding: `1.8591409142295225`

<details>
  <summary>SQL</summary>

```
-- Setup for evaluation strategy 'Dataflow rendering':
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, uint4_val UINT4);
INSERT INTO t_dfr_vert VALUES (7, 0::UINT4);
INSERT INTO t_dfr_vert VALUES (8, 1::UINT4);
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, uint4_val)
 AS SELECT 7, 0::UINT4
    UNION SELECT 8, 1::UINT4;
-- ------------------------------
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  avg(exp(uint4_val))
FROM
  t_dfr_vert
WHERE row_index IN (8, 7);
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  avg(exp(uint4_val))
FROM
  v_ctf_vert
WHERE row_index IN (8, 7);

```

</details>

#### Further ones
Happy to provide further examples if desired.

### Possibly related
* https://github.com/MaterializeInc/materialize/issues/19511
* https://github.com/MaterializeInc/materialize/issues/19577

### After the fix
Please remove entries from the ignore list in [inconsistency_ignore_filter.py](https://github.com/MaterializeInc/materialize/blob/main/misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py) or reach out to @nrainer-materialize.

**Comment by vmarcos** [comment]
@nrainer-materialize Some triaging of this issue:

- The trigonometric functions all take in `double precision` and produce `double precision`: https://materialize.com/docs/sql/functions/#trigonometric-func.
- So even though the source data is represented in integer types, we just have to cast it to floating point to apply the function.
- The output we get for the function application is the same in constant folding and in dataflow execution. The trouble is how we perform the `sum` subsequently.
- Since we have a `sum` of floating point numbers by the reasoning above, we lose precision by applying a `float_scale`. This is defined as https://github.com/MaterializeInc/materialize/blob/fd23c0f413b2f88c5722de881a6d784c4c181c41/src/compute/src/render/reduce.rs#L959
Then, the input value in the examples is converted as https://github.com/MaterializeInc/materialize/blob/fd23c0f413b2f88c5722de881a6d784c4c181c41/src/compute/src/render/reduce.rs#L1041-L1046
- I verified in one of your examples that this is the step where the precision is lost.

So in summary, this is unfortunately a consequence of how we re-cast floating point numbers into fixed point inside the rendering code. So these scenarios are further demonstrations of issue #15186.

I am closing this issue therefore as a duplicate; please feel free to reopen if you have other evidence.

**Comment by nrainer-materialize** [comment]
Thank you for the analysis!
